### PR TITLE
fix: Add OCIO configuration to Job Environments

### PR DESCRIPTION
**Pull request by diane-hj**
### What was the problem/requirement? (What/Why)
- There are 3 ways to set OCIO configuration in Nuke: environment variable, custom config, and default config that is defined in Nuke already. The original implementation only took care of custom config. We want to handle when customer set their OCIO configuration in environment variable or default config. 

### What was the solution? (How)
- When OCIO is enabled, check the configuration path for all three options, then set the path via JobEnvironment in JobTemplate. 
 
### What is the impact of this change?
- When environment variable is set, the path will be set with environment variable 
- When custom config is set, the path will be set with custom config
- When default config is set, the path will be set with default config 
 
### How was this change tested?
- Submitted the job with OCIO setup, and verified it rendered successfully 
- Submitted the above job with different OCIO setup, verified it rendered successfully, and verified the output looks different in color scheme 
- Submitted the job with OCIO environment variable, and verified it rendered successfully 

### Did you run the "Job Bundle Output Tests"? If not, why not? If so, paste the test results here.

```

Timestamp: 2024-09-16T22:51:08.146094-07:00
Running job bundle output test: /Users/chaejung/Desktop/submitter/deadline-cloud-for-nuke/job_bundle_output_tests/group-read

group-read
Test succeeded

Timestamp: 2024-09-16T22:51:08.578900-07:00
Running job bundle output test: /Users/chaejung/Desktop/submitter/deadline-cloud-for-nuke/job_bundle_output_tests/noise-saver

noise-saver
Test succeeded

Timestamp: 2024-09-16T22:51:08.968215-07:00
Running job bundle output test: /Users/chaejung/Desktop/submitter/deadline-cloud-for-nuke/job_bundle_output_tests/multi-load-save

multi-load-save
Test succeeded

Timestamp: 2024-09-16T22:51:09.358968-07:00
Running job bundle output test: /Users/chaejung/Desktop/submitter/deadline-cloud-for-nuke/job_bundle_output_tests/frame

frame
Test succeeded

Timestamp: 2024-09-16T22:51:09.757446-07:00
Running job bundle output test: /Users/chaejung/Desktop/submitter/deadline-cloud-for-nuke/job_bundle_output_tests/ocio

ocio
Test succeeded

Timestamp: 2024-09-16T22:51:10.157282-07:00
Running job bundle output test: /Users/chaejung/Desktop/submitter/deadline-cloud-for-nuke/job_bundle_output_tests/cwd-path

cwd-path
Test succeeded

All tests passed, ran 6 total.
Timestamp: 2024-09-16T22:51:13.578587-07:00

```

### Was this change documented?
No 

### Is this a breaking change?
No - In Nuke, environment variable takes precedence over other ocio configurations. (Ex. When OCIO environment variable is set, updating custom config in UI is disabled with the message: `OCIO environment variable is set. Setting OCIO configs from within Nuke will be disabled.`)

The existing user who uses custom config will continue to work. 

----

*By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.*

### CODE REVIEW

File: job_bundle_output_tests/ocio/expected_job_bundle/template.yaml
Comment (eherozhao) [inline job_bundle_output_tests/ocio/expected_job_bundle/template.yaml]: nit: I think we don't need to have `\n`
It can be 
```
...
    runnable: true
    data: | 
        #!/bin/bash
        echo 'openjd_env: OCIO={{Param.OCIOConfigPath}}'
```

File: src/deadline/nuke_submitter/assets.py
Comment (crowecawcaw) [inline src/deadline/nuke_submitter/assets.py:88]: Is the ocio config path relative and not absolute? Are there any issues with path mapping or anything like that?

File: src/deadline/nuke_submitter/assets.py
Comment (crowecawcaw) [inline src/deadline/nuke_submitter/assets.py]: Nit: remove

File: src/deadline/nuke_submitter/assets.py
Comment (crowecawcaw) [inline src/deadline/nuke_submitter/assets.py]: Nit: maybe cleaner if we just return the first discovered OCIO config instead of nesting ifs

File: src/deadline/nuke_submitter/deadline_submitter_for_nuke.py
Comment (crowecawcaw) [inline src/deadline/nuke_submitter/deadline_submitter_for_nuke.py]: Should we check if the path is `None`?

File: src/deadline/nuke_submitter/assets.py
Comment (crowecawcaw) [inline src/deadline/nuke_submitter/assets.py]: Should this error block the submission? What happens if it's missing?

File: src/deadline/nuke_submitter/assets.py
Comment (diane-hj) [reply src/deadline/nuke_submitter/assets.py:88]: ocio config path is in absolute. `get_ocio_config_path()` will return absolute path for stock and custom configurations. Yeah, when testing, it didn't work properly with relative path. :) Please let me know if you have any concerns.

File: src/deadline/nuke_submitter/deadline_submitter_for_nuke.py
Comment (crowecawcaw) [inline src/deadline/nuke_submitter/deadline_submitter_for_nuke.py]: I think we can define environment variables directly in OJD instead of with a script: https://github.com/OpenJobDescription/openjd-specifications/wiki/2023-09-Template-Schemas#4-environment

I haven't used it personally, but it looks like it'd do what we want.

File: src/deadline/nuke_util/ocio.py
Comment (crowecawcaw) [inline src/deadline/nuke_util/ocio.py]: Comment needs to be updated

**Review by crowecawcaw — CHANGES_REQUESTED** [review]

File: src/deadline/nuke_submitter/assets.py
Comment (diane-hj) [reply src/deadline/nuke_submitter/assets.py]: Dang! I thought I erased all of it. haha Thanks for catching it.

File: src/deadline/nuke_submitter/deadline_submitter_for_nuke.py
Comment (diane-hj) [reply src/deadline/nuke_submitter/deadline_submitter_for_nuke.py]: Nice! Thanks, its much simpler. 🙏

File: src/deadline/nuke_submitter/assets.py
Comment (diane-hj) [reply src/deadline/nuke_submitter/assets.py]: If its missing, it means that OCIO is enabled but the provided file is not an existing file or working file. In this case, it will not open Submitter pop up and instead will show the alert message. 
Screenshot 2024-09-17 at 11 29 59 AM

Comment (diane-hj) [reply src/deadline/nuke_submitter/assets.py]: Actually, I take this back. It will still show the Submitter pop up after this alert. I will make changes to not show the submitter if the OCIO config file is not specified properly.

File: job_bundle_output_tests/ocio/expected_job_bundle/template.yaml
Comment (diane-hj) [reply job_bundle_output_tests/ocio/expected_job_bundle/template.yaml]: This block is deleted :).

File: src/deadline/nuke_submitter/assets.py
Comment (diane-hj) [reply src/deadline/nuke_submitter/assets.py]: Updated. Please let me know if I misunderstood your comment 🙏

File: src/deadline/nuke_submitter/deadline_submitter_for_nuke.py
Comment (diane-hj) [reply src/deadline/nuke_submitter/deadline_submitter_for_nuke.py]: Although it should not happen when OCIO is enabled, I added safe guard for this just in case, somehow it happens, customers would know what they should do to unblock themselves. Thanks!

**Review by crowecawcaw — DISMISSED** [review]

File: src/deadline/nuke_submitter/deadline_submitter_for_nuke.py
Comment (lucaseck) [inline src/deadline/nuke_submitter/deadline_submitter_for_nuke.py:179]: What's the use case for removing it? Is the OCIO path already there in the template, I thought it was only added if OCIO was enabled two lines above. Can the param only be added if OCIO is being used and the default template left the same?

File: src/deadline/nuke_submitter/default_nuke_job_template.yaml
Comment (lucaseck) [inline src/deadline/nuke_submitter/default_nuke_job_template.yaml]: This looks to be just a single FILE not a directory like the description says. One or the other probably needs to be updated.

File: job_bundle_output_tests/ocio/expected_job_bundle/template.yaml
Comment (lucaseck) [reply job_bundle_output_tests/ocio/expected_job_bundle/template.yaml:144]: I might be missing some context, this block still looks like it exists and has the `\n` everywhere. Is that just a consequence of the environment being written out or do we have control over what this block looks like in the generated template? If we have control then personally I prefer the way that Hao outlined.

File: src/deadline/nuke_submitter/deadline_submitter_for_nuke.py
Comment (diane-hj) [reply src/deadline/nuke_submitter/deadline_submitter_for_nuke.py:179]: The remove method is to remove any OCIO information from the job_template's parameter definitions. If there is, I think it will complain about not having the default value, etc. Also, I thought if its not being used, it is ideal to remove it from the parameter definition. Please let me know how you think about it 🙏 

The upper part of adding is to add the `jobEnvironment` definition for it using the parameter definition that is in default job template. Please let me know if it makes sense!

File: src/deadline/nuke_submitter/default_nuke_job_template.yaml
Comment (diane-hj) [reply src/deadline/nuke_submitter/default_nuke_job_template.yaml]: Thanks for catching, I will update this to something like, 
`description: The OCIO config file used by this job`

File: job_bundle_output_tests/ocio/expected_job_bundle/template.yaml
Comment (diane-hj) [reply job_bundle_output_tests/ocio/expected_job_bundle/template.yaml]: Oh no, this part was updated but not included in the commit. I will add it in the next commit. This was updated to something like 
```
jobEnvironments:
- name: Add OCIO Path to Environment Variable
  variables:
    OCIO: '{{Param.OCIOConfigPath}}'
```

File: src/deadline/nuke_submitter/deadline_submitter_for_nuke.py
Comment (lucaseck) [reply src/deadline/nuke_submitter/deadline_submitter_for_nuke.py:179]: Can we remove the parameter entirely from the default job template though? Only add the parameter into the template if OCIO is enabled. Then the remove method wouldn't be needed at all and OCIO relevant stuff would only be included if enabled

Comment (lucaseck) [reply src/deadline/nuke_submitter/deadline_submitter_for_nuke.py:179]: Spoke offline, personally I'm of the opinion that the parameter without the environment and vice versa is not really useful so I don't see why we would bake one into the default template, but not the other. Given Gizmo handling already follows this pattern though I'm not going to block the review on it, but leave it up to the owners to decide whether both OCIO and Gizmo parameters should be in the default template or not.

File: src/deadline/nuke_submitter/default_nuke_job_template.yaml
Comment (lucaseck) [inline src/deadline/nuke_submitter/default_nuke_job_template.yaml:32]: Is there a default that can make sense here?

Comment (diane-hj) [reply src/deadline/nuke_submitter/default_nuke_job_template.yaml:32]: Someone else can correct me if I'm wrong. 🙏 
I don't think we have a perfect default path T-T because it's really up to customers. Customers can bring their configuration by themselves or use Nuke's default configurations where Nuke's default has more than 3 configuration that customers can choose so it is hard to choose one.

Comment (lucaseck) [reply src/deadline/nuke_submitter/default_nuke_job_template.yaml:32]: Makes sense 👍

**Review by lucaseck — APPROVED** [review]

**Review by crowecawcaw — APPROVED** [review]

**Pull request merged**
